File: implementations/java/client/src/main/java/com/theshuai/tunnelclient/peer/PeerIpPacket.java
package com.theshuai.tunnelclient.peer;

final class PeerIpPacket {
    private static final int IPV4_MIN_HEADER_BYTES = 20;
    private static final int IPV4_PROTOCOL_ICMP = 1;
    private static final int IPV4_PROTOCOL_TCP = 6;
    private static final int IPV4_PROTOCOL_UDP = 17;
    private static final int ICMP_ECHO_REPLY = 0;
    private static final int ICMP_ECHO_REQUEST = 8;

    private PeerIpPacket() {
    }

    static String destinationIpv4(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return "";
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        if (version != 4 || ihl < 5 || packet.length < ihl * 4) {
            return "";
        }
        return (packet[16] & 0xFF) + "."
                + (packet[17] & 0xFF) + "."
                + (packet[18] & 0xFF) + "."
                + (packet[19] & 0xFF);
    }

    static String sourceIpv4(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return "";
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        if (version != 4 || ihl < 5 || packet.length < ihl * 4) {
            return "";
        }
        return (packet[12] & 0xFF) + "."
                + (packet[13] & 0xFF) + "."
                + (packet[14] & 0xFF) + "."
                + (packet[15] & 0xFF);
    }

    static byte[] icmpEchoReplyFor(byte[] packet, String localVirtualIp) {
        if (!isIcmpEchoRequestFor(packet, localVirtualIp)) {
            return null;
        }
        int ihlBytes = (packet[0] & 0x0F) * 4;
        int totalLength = totalLength(packet);
        byte[] reply = new byte[totalLength];
        System.arraycopy(packet, 0, reply, 0, totalLength);

        for (int i = 0; i < 4; i++) {
            byte source = reply[12 + i];
            reply[12 + i] = reply[16 + i];
            reply[16 + i] = source;
        }
        reply[8] = 64;
        reply[10] = 0;
        reply[11] = 0;

        int icmpOffset = ihlBytes;
        reply[icmpOffset] = ICMP_ECHO_REPLY;
        reply[icmpOffset + 1] = 0;
        reply[icmpOffset + 2] = 0;
        reply[icmpOffset + 3] = 0;
        int icmpChecksum = checksum(reply, icmpOffset, totalLength - icmpOffset);
        reply[icmpOffset + 2] = (byte) ((icmpChecksum >>> 8) & 0xFF);
        reply[icmpOffset + 3] = (byte) (icmpChecksum & 0xFF);

        int ipChecksum = checksum(reply, 0, ihlBytes);
        reply[10] = (byte) ((ipChecksum >>> 8) & 0xFF);
        reply[11] = (byte) (ipChecksum & 0xFF);
        return reply;
    }

    static int protocol(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return -1;
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        if (version != 4 || ihl < 5 || packet.length < ihl * 4) {
            return -1;
        }
        return packet[9] & 0xFF;
    }

    static String describe(byte[] packet) {
        int protocol = protocol(packet);
        String protocolName = switch (protocol) {
            case IPV4_PROTOCOL_ICMP -> "ICMP";
            case IPV4_PROTOCOL_TCP -> "TCP";
            case IPV4_PROTOCOL_UDP -> "UDP";
            default -> protocol < 0 ? "IPv4?" : "IP-" + protocol;
        };
        String source = sourceIpv4(packet);
        String target = destinationIpv4(packet);
        int sourcePort = sourcePort(packet);
        int targetPort = destinationPort(packet);
        if ((protocol == IPV4_PROTOCOL_TCP || protocol == IPV4_PROTOCOL_UDP) && sourcePort > 0 && targetPort > 0) {
            return protocolName + " " + source + ":" + sourcePort + " -> " + target + ":" + targetPort;
        }
        return protocolName + " " + source + " -> " + target;
    }

    static String flowKey(byte[] packet) {
        return protocol(packet) + "|" + sourceIpv4(packet) + "|" + sourcePort(packet)
                + "|" + destinationIpv4(packet) + "|" + destinationPort(packet);
    }

    private static boolean isIcmpEchoRequestFor(byte[] packet, String localVirtualIp) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES || !destinationIpv4(packet).equals(localVirtualIp)) {
            return false;
        }
        int ihlBytes = (packet[0] & 0x0F) * 4;
        int totalLength = totalLength(packet);
        return totalLength >= ihlBytes + 8
                && totalLength <= packet.length
                && (packet[9] & 0xFF) == IPV4_PROTOCOL_ICMP
                && (packet[ihlBytes] & 0xFF) == ICMP_ECHO_REQUEST
                && (packet[ihlBytes + 1] & 0xFF) == 0;
    }

    private static int totalLength(byte[] packet) {
        if (packet == null || packet.length < 4) {
            return 0;
        }
        int totalLength = ((packet[2] & 0xFF) << 8) | (packet[3] & 0xFF);
        return totalLength > 0 && totalLength <= packet.length ? totalLength : packet.length;
    }

    private static int sourcePort(byte[] packet) {
        int offset = transportOffset(packet);
        if (offset < 0 || packet.length < offset + 4) {
            return -1;
        }
        int protocol = protocol(packet);
        if (protocol != IPV4_PROTOCOL_TCP && protocol != IPV4_PROTOCOL_UDP) {
            return -1;
        }
        return ((packet[offset] & 0xFF) << 8) | (packet[offset + 1] & 0xFF);
    }

    private static int destinationPort(byte[] packet) {
        int offset = transportOffset(packet);
        if (offset < 0 || packet.length < offset + 4) {
            return -1;
        }
        int protocol = protocol(packet);
        if (protocol != IPV4_PROTOCOL_TCP && protocol != IPV4_PROTOCOL_UDP) {
            return -1;
        }
        return ((packet[offset + 2] & 0xFF) << 8) | (packet[offset + 3] & 0xFF);
    }

    private static int transportOffset(byte[] packet) {
        if (packet == null || packet.length < IPV4_MIN_HEADER_BYTES) {
            return -1;
        }
        int version = (packet[0] >>> 4) & 0x0F;
        int ihl = packet[0] & 0x0F;
        int offset = ihl * 4;
        if (version != 4 || ihl < 5 || packet.length < offset) {
            return -1;
        }
        return offset;
    }

    private static int checksum(byte[] packet, int offset, int length) {
        long sum = 0;
        int index = offset;
        int end = offset + length;
        while (index + 1 < end) {
            sum += ((packet[index] & 0xFF) << 8) | (packet[index + 1] & 0xFF);
            index += 2;
        }
        if (index < end) {
            sum += (packet[index] & 0xFF) << 8;
        }
        while ((sum >>> 16) != 0) {
            sum = (sum & 0xFFFF) + (sum >>> 16);
        }
        return (int) (~sum) & 0xFFFF;
    }
}

File: implementations/csharp/client/src/ShuaiTunnel.Client/PeerMesh/PeerIpPacket.cs
using System.Buffers.Binary;
using System.Net;

namespace ShuaiTunnel.Client.PeerMesh;

internal static class PeerIpPacket
{
    private const byte ProtocolIcmp = 1;
    private const byte ProtocolTcp = 6;
    private const byte ProtocolUdp = 17;
    private const byte IcmpEchoReply = 0;
    private const byte IcmpEchoRequest = 8;

    public static string DestinationIPv4(ReadOnlySpan<byte> packet)
    {
        if (!IsIPv4(packet))
        {
            return "";
        }
        return new IPAddress(packet.Slice(16, 4)).ToString();
    }

    public static string SourceIPv4(ReadOnlySpan<byte> packet)
    {
        if (!IsIPv4(packet))
        {
            return "";
        }
        return new IPAddress(packet.Slice(12, 4)).ToString();
    }

    public static string FlowKey(ReadOnlySpan<byte> packet)
    {
        var source = SourceIPv4(packet);
        var target = DestinationIPv4(packet);
        if (source.Length == 0 || target.Length == 0)
        {
            return "";
        }
        var protocol = packet[9];
        var ihl = (packet[0] & 0x0f) * 4;
        var sourcePort = 0;
        var targetPort = 0;
        if ((protocol is ProtocolTcp or ProtocolUdp) && packet.Length >= ihl + 4)
        {
            sourcePort = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(ihl, 2));
            targetPort = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(ihl + 2, 2));
        }
        return $"{source}:{sourcePort}->{target}:{targetPort}/{protocol}";
    }

    public static byte[]? IcmpEchoReplyFor(ReadOnlySpan<byte> packet, string? localVirtualIp)
    {
        if (!IsIcmpEchoRequestFor(packet, localVirtualIp))
        {
            return null;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        var totalLength = TotalLength(packet);
        var reply = packet[..totalLength].ToArray();
        reply.AsSpan(16, 4).CopyTo(reply.AsSpan(12, 4));
        packet.Slice(12, 4).CopyTo(reply.AsSpan(16, 4));
        reply[8] = 64;
        reply[10] = 0;
        reply[11] = 0;

        reply[ihl] = IcmpEchoReply;
        reply[ihl + 1] = 0;
        reply[ihl + 2] = 0;
        reply[ihl + 3] = 0;
        BinaryPrimitives.WriteUInt16BigEndian(reply.AsSpan(ihl + 2, 2), Checksum(reply.AsSpan(ihl, totalLength - ihl)));
        BinaryPrimitives.WriteUInt16BigEndian(reply.AsSpan(10, 2), Checksum(reply.AsSpan(0, ihl)));
        return reply;
    }

    private static bool IsIPv4(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < 20 || packet[0] >> 4 != 4)
        {
            return false;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        return ihl >= 20 && packet.Length >= ihl;
    }

    private static bool IsIcmpEchoRequestFor(ReadOnlySpan<byte> packet, string? localVirtualIp)
    {
        if (DestinationIPv4(packet) != localVirtualIp)
        {
            return false;
        }
        var ihl = (packet[0] & 0x0f) * 4;
        var totalLength = TotalLength(packet);
        return totalLength >= ihl + 8
            && totalLength <= packet.Length
            && packet[9] == ProtocolIcmp
            && packet[ihl] == IcmpEchoRequest
            && packet[ihl + 1] == 0;
    }

    private static int TotalLength(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < 4)
        {
            return 0;
        }
        var totalLength = BinaryPrimitives.ReadUInt16BigEndian(packet.Slice(2, 2));
        return totalLength > 0 && totalLength <= packet.Length ? totalLength : packet.Length;
    }

    internal static ushort Checksum(ReadOnlySpan<byte> data)
    {
        uint sum = 0;
        var index = 0;
        while (index + 1 < data.Length)
        {
            sum += BinaryPrimitives.ReadUInt16BigEndian(data.Slice(index, 2));
            index += 2;
        }
        if (index < data.Length)
        {
            sum += (uint)data[index] << 8;
        }
        while ((sum >> 16) != 0)
        {
            sum = (sum & 0xffff) + (sum >> 16);
        }
        return (ushort)~sum;
    }
}
